<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API &mdash; variantpost 0.0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial" href="tutorial.html" />
    <link rel="prev" title="variantPost: Python library for small variants" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            variantpost
          </a>
              <div class="version">
                0.0.1
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#variant-processing-on-the-reference">variant processing on the reference</a></li>
<li class="toctree-l2"><a class="reference internal" href="#variant-processing-in-alignment-file">variant processing in alignment file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">variantpost</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">API</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="api">
<h1>API<a class="headerlink" href="#api" title="Permalink to this heading"></a></h1>
<section id="variant-processing-on-the-reference">
<h2>variant processing on the reference<a class="headerlink" href="#variant-processing-on-the-reference" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="variantpost.Variant">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">variantpost.</span></span><span class="sig-name descname"><span class="pre">Variant</span></span><a class="headerlink" href="#variantpost.Variant" title="Permalink to this definition"></a></dt>
<dd><p>This class abstracts a variant relative to a linear reference genome.
Equality holds between <a class="reference internal" href="#variantpost.Variant" title="variantpost.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> objects
if they are identical in the normalized form (equivalent alignments)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<em>string</em>) – chromosome name.</p></li>
<li><p><strong>pos</strong> (<em>integer</em>) – 1-based genomic position.</p></li>
<li><p><strong>ref</strong> (<em>string</em>) – VCF-style reference allele.</p></li>
<li><p><strong>alt</strong> (<em>string</em>) – VCF-style alternative allele.</p></li>
<li><p><strong>reference</strong> (<em>pysam.Fasta</em>) – reference FASTA file supplied as
<a class="reference external" href="https://pysam.readthedocs.io/en/latest/api.html#pysam.FastaFile">pysam.FastaFile</a> object.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – if the input alleles contain letters other than A, C, G, T and N.</p>
</dd>
</dl>
<dl class="py property">
<dt class="sig sig-object py" id="variantpost.Variant.is_del">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">is_del</span></span><a class="headerlink" href="#variantpost.Variant.is_del" title="Permalink to this definition"></a></dt>
<dd><p>True for deletions. False for insertions or substitutions.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="variantpost.Variant.is_indel">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">is_indel</span></span><a class="headerlink" href="#variantpost.Variant.is_indel" title="Permalink to this definition"></a></dt>
<dd><p>True for insertions or deletions. False otherwise.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="variantpost.Variant.is_ins">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">is_ins</span></span><a class="headerlink" href="#variantpost.Variant.is_ins" title="Permalink to this definition"></a></dt>
<dd><p>True for insetions. False for deletions or substitutions.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="variantpost.Variant.is_normalized">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">is_normalized</span></span><a class="headerlink" href="#variantpost.Variant.is_normalized" title="Permalink to this definition"></a></dt>
<dd><p>True if left-aligened and the allele representations are minimal.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="variantpost.Variant.is_simple_indel">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">is_simple_indel</span></span><a class="headerlink" href="#variantpost.Variant.is_simple_indel" title="Permalink to this definition"></a></dt>
<dd><p>True for indel that is not complex (co-occurrence of insertion and deletion).
False for complex indels or substitutions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="variantpost.Variant.left_pos">
<span class="sig-name descname"><span class="pre">left_pos</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#variantpost.Variant.left_pos" title="Permalink to this definition"></a></dt>
<dd><p>returns a left-aligned genomic position.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="variantpost.Variant.normalize">
<span class="sig-name descname"><span class="pre">normalize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#variantpost.Variant.normalize" title="Permalink to this definition"></a></dt>
<dd><p>left-aligns in genomic position and minimize the allele represention.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>inplace</strong> (<em>bool</em>) – normalizes this (self) <a class="reference internal" href="#variantpost.Variant" title="variantpost.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> object if True.
Otherwise, returns a normalized copy of the object. Default to False.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="variantpost.Variant.query_vcf">
<span class="sig-name descname"><span class="pre">query_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vcf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">match_by_equivalence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#variantpost.Variant.query_vcf" title="Permalink to this definition"></a></dt>
<dd><p>returns a <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list">list</a> of <code class="xref py py-class docutils literal notranslate"><span class="pre">MatchedRecord</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vcf</strong> (<em>pysam.VariantFile</em>) – VCF file to be queried.
Supply as <a class="reference external" href="https://pysam.readthedocs.io/en/latest/api.html#pysam.VariantFile">pysam.VariantFile</a> object. The VCF file muet be indexed.</p></li>
<li><p><strong>chrom_name</strong> (<em>string</em>) – specify an alias chromosome name if the VCF file uses a chromosome nomenclature different from the reference genome in <a class="reference internal" href="#variantpost.Variant" title="variantpost.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a>.
If not specified (default), the nomenclature in the FASTA file in <a class="reference internal" href="#variantpost.Variant" title="variantpost.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> will be used.</p></li>
<li><p><strong>match_by_equivalence</strong> (<em>bool</em>) – queries the VCF records by normalization if True (default). Otherwise, positionally overlapping records will be returned.</p></li>
</ul>
</dd>
</dl>
<p><code class="xref py py-class docutils literal notranslate"><span class="pre">MatchedRecord</span></code> is a <a class="reference external" href="https://docs.python.org/3/library/collections.html#collections.namedtuple">namedtuple</a>
with the following fields</p>
<ul class="simple">
<li><p><strong>chrom</strong> - VCF CHROM field.</p></li>
<li><p><strong>pos</strong> - VCF POS field.</p></li>
<li><p><strong>id</strong> - VCF ID field.</p></li>
<li><p><strong>ref</strong> - VCF REF field.</p></li>
<li><p><strong>alts</strong> - VCF ALT field as <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuples">tuple</a>. May contain multiple alleles.</p></li>
<li><p><strong>qual</strong> - VCF QUAL field.</p></li>
<li><p><strong>filter</strong> - VCF FILTER field.</p></li>
<li><p><strong>info</strong> - VCF INFO field. Values are accessible by keys defined in the header.</p></li>
<li><p><strong>format</strong> - VCF FORMAT field.</p></li>
<li><p><strong>samples</strong> - VCF genotype field. Genotypes are accessible by using sample names as key.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="variantpost.Variant.right_pos">
<span class="sig-name descname"><span class="pre">right_pos</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#variantpost.Variant.right_pos" title="Permalink to this definition"></a></dt>
<dd><p>returns the variant-end position after right-aligned.</p>
</dd></dl>

</dd></dl>

</section>
<section id="variant-processing-in-alignment-file">
<h2>variant processing in alignment file<a class="headerlink" href="#variant-processing-in-alignment-file" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="variantpost.VariantAlignment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">variantpost.</span></span><span class="sig-name descname"><span class="pre">VariantAlignment</span></span><a class="headerlink" href="#variantpost.VariantAlignment" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>This class accepts the target variant as <a class="reference internal" href="#variantpost.Variant" title="variantpost.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> and the BAM file</dt><dd><p>as <a class="reference external" href="https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignmentFile">pysam.AlignmentFile</a>
to process the variant alignment.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant</strong> (<a class="reference internal" href="#variantpost.Variant" title="variantpost.Variant"><em>Variant</em></a>) – <a class="reference internal" href="#variantpost.Variant" title="variantpost.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> object representing the target variant.</p></li>
<li><p><strong>bam</strong> (<em>pysam.AlignmentFile</em>) – BAM file supplied as
<a class="reference external" href="https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignmentFile">pysam.AlignmentFile</a> object.</p></li>
<li><p><strong>second_bam</strong> (<em>pysam.AlignmentFile</em>) – A second BAM file for paired analysis. Default: None.</p></li>
<li><p><strong>chrom_name</strong> (<em>string</em>) – Specify an alias chromosome name if the BAM file uses a chromosome nomenclature different from
the reference used in <a class="reference internal" href="#variantpost.Variant" title="variantpost.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a>. If not specified (default), the nomenclature
in the reference will be used.</p></li>
<li><p><strong>mapping_quality_threshold</strong> (<em>integer</em>) – A mininum mapping quality to be analized. Default 1.</p></li>
<li><p><strong>base_quality_threshold</strong> (<em>integer</em>) – Non-reference base-calls with a Phred-scale quality score below the threshold are labeled low quality.</p></li>
<li><p><strong>low_quality_base_rate_threshold</strong> (<em>float</em>) – </p></li>
<li><p><strong>downsample_threshold</strong> (<em>integer</em>) – </p></li>
<li><p><strong>match_score</strong> (<em>integer</em>) – </p></li>
<li><p><strong>mismatch_penalty</strong> (<em>integer</em>) – </p></li>
<li><p><strong>gap_open_penalty</strong> (<em>integer</em>) – </p></li>
<li><p><strong>gap_extension_penalty</strong> (<em>integer</em>) – </p></li>
<li><p><strong>kmer_size</strong> (<em>integer</em>) – </p></li>
<li><p><strong>local_threshold</strong> (<em>integer</em>) – </p></li>
<li><p><strong>match_penalty_for_phasing</strong> (<em>float</em>) – </p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="variantpost.VariantAlignment.count_alleles">
<span class="sig-name descname"><span class="pre">count_alleles</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#variantpost.VariantAlignment.count_alleles" title="Permalink to this definition"></a></dt>
<dd><p>returns <code class="xref py py-class docutils literal notranslate"><span class="pre">AlleleCount</span></code> as <a class="reference external" href="https://docs.python.org/3/library/collections.html#collections.namedtuple">namedtuple</a> of read counts.
<code class="xref py py-class docutils literal notranslate"><span class="pre">AlleleCount</span></code> has the following fields accessible by attribute:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>s</strong> - count of read names supporting the variant.</p></li>
<li><p><strong>n</strong> - count of read names not supporting the variant.</p></li>
<li><p><strong>u</strong> - count of read names undetermined to be supporting/non-supporting</p></li>
</ul>
<p>Strand breakdowns are also available by:</p>
<ul class="simple">
<li><p><strong>s_fw</strong> - count of forward reads supportingt the variant.</p></li>
<li><p><strong>s_rv</strong> - count of reverse reads supportingt the variant.</p></li>
<li><p>…</p></li>
</ul>
<p>For paired analysis, <code class="xref py py-class docutils literal notranslate"><span class="pre">PairedAlleleCount</span></code> is returnd and has the following fields:</p>
<ul class="simple">
<li><p><strong>first</strong> - <code class="xref py py-class docutils literal notranslate"><span class="pre">AlleleCount</span></code> for the first BAM file.</p></li>
<li><p><strong>second</strong> - <code class="xref py py-class docutils literal notranslate"><span class="pre">AlleleCount</span></code> for the second BAM file.</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="variantpost.VariantAlignment.phase">
<span class="sig-name descname"><span class="pre">phase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match_penalty_for_phasing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_common_substr_len</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#variantpost.VariantAlignment.phase" title="Permalink to this definition"></a></dt>
<dd><p>returns <a class="reference internal" href="#variantpost.Variant" title="variantpost.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> representing a phased target variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>max_common_substr_len</strong> (<em>int</em>) – </p></li>
<li><p><strong>match_penalty_for_phasing</strong> (<em>float</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="variantPost: Python library for small variants" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorial.html" class="btn btn-neutral float-right" title="Tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Kohei Hagiwara.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>